doc/userguide/capture-hardware/ebpf-xdp.rst

```diff
@@ -350,11 +365,129 @@ Confirm you have the XDP filter engaged in the output (example)::
+for that will be used to check for the presence of the XDP filter ::
+
+  - interface: eth3
+    pinned-maps: true
+    pinned-maps-name: ipv4_drop
+    xdp-filter-file: /etc/suricata/ebpf/xdp_filter.bpf
+
+If XDP bypass is used in IPS mode stopping Suricata will trigger a interruption in the traffic.
+To fix that, the provided XDP filter `xdp_filter.bpf` is containing a map that will trigger
+a global bypass if set to 1. You need to use `pinned-maps` to benefit of this feature.
+
+To use it you need to set `#define USE_GLOBAL_BYPASS   1` (instead of 0) in the `xdp_filter.c` file and rebuild
+the eBPF code and install the eBPF file in the correct place. If you write `1` as key `0` then the XDP
+filter will switch to global bypass mode. Set key `0` to `0` to send traffic to Suricata.
+
+The switch must be activated on all sniffing interfaces. For an interface named `eth0` the global
+switch map will be `/sys/fs/bpf/suricata-eth0-global_bypass`.
+
+Hardware bypass with Netronome
+------------------------------
+
+Netronome card supports hardware bypass. In this case the eBPF code is running in the card
+itself. This introduces some architectural differences compared to driver mode and the configuration
+and eBPF filter need to be updated.
+
+On eBPF side, as of Linux 4.19 CPU maps and interfaces redirect are not supported and these features
+need to be disabled. By architecture, per CPU hash should not be used and have to be disabled.
+To achieve this, edit the beginning of `ebpf/xdp_filter.c` and do ::
+
+ #define BUILD_CPUMAP        0
+ /* Increase CPUMAP_MAX_CPUS if ever you have more than 64 CPUs */
+ #define CPUMAP_MAX_CPUS     64
+
+ #define USE_PERCPU_HASH    0
+ #define GOT_TX_PEER    0
+
+Then build the bpf file with `make` and install it in the expected place.
+
+On Suricata configuration side, this is rather simple as you need to activate
+hardware mode and the `no-percpu-hash` option in the `af-packet` configuration
+of the interface ::
+
+    xdp-mode: hw
+    no-percpu-hash: true
+
+The load  balancing will be done on IP pairs inside the eBPF code, so
+using `cluster_qm` as cluster type is a good idea ::
+
+    cluster-type: cluster_qm
+
+As of Linux 4.19, the number of threads must be a power of 2. So set
+`threads` variable interface of the `af-packet` interface to a power
+of 2 and in the eBPF filter set the following variable accordingly ::
+
+ #define RSS_QUEUE_NUMBERS   32
 
 Getting live info about bypass
 ------------------------------
 
 You can get information about bypass via the stats event and through the unix socket.
-`ìface-stat` will return the number of bypassed packets (adding packets for a flow when it timeout).
-`ebpf-bypassed-stats` command will return the number of element in IPv4 and IPv6 flow tables for
-each interfaces.
+``iface-stat`` will return the number of bypassed packets (adding packets for a flow when it timeout) ::
```

Comment:
Only way I see to do that is build a number when we dump the IPv4 and IPv6 flow table. We could take the data compare to previous run for each individual flow and add this to the interface at the end of the table dump. That could work.

Reply:
This appear to be trickier than thought. Algorithm is the following:
- at bypass time we create entries in the map with counter from the flow
- eBPF code update the counter for flow as packets are seen
- at regular interval we dump the flow tables, and sync Suricata counter with eBPF map counter

There is a problem with the packets of a flow that have just been bypassed. In a high percentage of the case, there is packets for a flow in the ring buffer at the moment we decide to bypass the flow. This packet can not be accounted by the eBPF code and are (with current code) accounted inside Suricata. This means we get out of sync. At first dump of the eBPF map with this flow contained, the counter are synchronized but the transition state is out of sync.

We could add a specific counter to the flow for packets received by a bypassed flow and then do a sync but it appears to be kind of overkill.

@victorjulien what do you think ?

Reply:
Two proposals:
- one with incorrect accounting: https://github.com/regit/suricata/commit/cab4b831a1942bbf67cc97dbe997cde741811d06
- one expensive but with correct accounting and more information: https://github.com/regit/suricata/commit/d8994a62d21683e283a2202849c68156492fc514

I like the second one because we have accurate accounting

Reply:
Hmm I'm failing to understand the problem, but both suggested patches seem not ideal. The 2nd may be better for bypass, but increase cost for all other ways of using Suricata.

Reply:
Another way to implement 2 would be to use a flow storage for the bypass stats. But this would require allocation for every bypass action and it will cost at least a pointer size.